File: packages/skill-eval-python/src/skill_eval/schemas/evals.py
from __future__ import annotations

from pydantic import BaseModel, Field, model_validator


class EvalCase(BaseModel):
    id: int
    prompt: str | None = None
    turns: list[str] | None = None
    expected_output: str
    files: list[str] = Field(default_factory=list)
    assertions: list[str] = Field(min_length=1)

    @model_validator(mode="after")
    def prompt_or_turns_required(self) -> EvalCase:
        if not self.prompt and not self.turns:
            raise ValueError("Either 'prompt' or 'turns' must be provided")
        if self.turns is not None and len(self.turns) < 1:
            raise ValueError("'turns' must have at least one element")
        return self

    @model_validator(mode="after")
    def prompt_and_turns_exclusive(self) -> EvalCase:
        if self.prompt and self.turns:
            raise ValueError(
                "'prompt' and 'turns' are mutually exclusive — use one or the other"
            )
        return self


class EvalsFile(BaseModel):
    skill_name: str
    models: list[str] | None = None
    evals: list[EvalCase] = Field(min_length=1)


def get_turns(eval_case: EvalCase) -> list[str]:
    if eval_case.turns:
        return eval_case.turns
    if eval_case.prompt:
        return [eval_case.prompt]
    return []

File: packages/skillet/src/eval/schemas/evals.ts
import { z } from "zod";
import { MockConfigSchema } from "../config/schema.js";

export const EvalMockScenarioSchema = z.object({
	state: z.record(z.unknown()).default({}),
	overrides: z
		.record(
			z.object({
				response: z.unknown().optional(),
				responseFromState: z.string().optional(),
			}),
		)
		.default({}),
});

export const InlineMockSchema = MockConfigSchema.extend({ name: z.string() });
export type InlineMock = z.infer<typeof InlineMockSchema>;

export const EvalCaseSchema = z
	.object({
		id: z.number(),
		prompt: z.string().optional(),
		turns: z.array(z.string()).min(1).optional(),
		expected_output: z.string(),
		files: z.array(z.string()).default([]),
		mocks: z.record(EvalMockScenarioSchema).default({}),
		assertions: z.array(z.string()).min(1),
	})
	.refine((v) => v.prompt || v.turns, {
		message: "Either 'prompt' or 'turns' must be provided",
	})
	.refine((v) => !(v.prompt && v.turns), {
		message: "'prompt' and 'turns' are mutually exclusive — use one or the other",
	});

export const EvalsFileSchema = z.object({
	skill_name: z.string(),
	models: z.array(z.string()).optional(),
	providers: z.array(z.string()).optional(),
	mocks: z.array(z.union([z.string(), InlineMockSchema])).optional(),
	evals: z.array(EvalCaseSchema).min(1),
});

export type EvalCase = z.infer<typeof EvalCaseSchema>;
export type EvalsFile = z.infer<typeof EvalsFileSchema>;
export type EvalMockScenario = z.infer<typeof EvalMockScenarioSchema>;

/** Normalize to a turns array regardless of whether prompt or turns was used. */
export function getTurns(evalCase: EvalCase): string[] {
	if (evalCase.turns) return evalCase.turns;
	if (evalCase.prompt) return [evalCase.prompt];
	return [];
}
